### ISSUE TITLE
compile failed for tf-gpu 1.4, cuda 9, cudnn 7, vc 2017, windows 10.

### ISSUE THREAD

**Issue by colmantse** [tensorflow/tensorflow#14691]
I would greatly appreciate if anyone could help me out with compiling.
I followed #5600 and #13962 in compiling the wheel, however when building i got into 6 types of in total 90 errors. Respectively c2070, c2059, c2064, c2001, c1057, c2146.

My build command was:
`cmake .. -A x64 -DCMAKE_BUILD_TYPE=Release ^
-DSWIG_EXECUTABLE=C:/swigwin-3.0.12/swig.exe ^
-DPYTHON_EXECUTABLE=python.exe ^
-DPYTHON_LIBRARIES=C:/ProgramData/Anaconda3/libs/python36_lib ^
-DPYTHON_INCLUDE_DIR=C:/ProgramData/Anaconda3/Include ^
-DNUMPY_INCLUDE_DIR=C:/ProgramData/Anaconda3/lib/site-packages/numpy/core/include ^
-Dtensorflow_ENABLE_GPU=ON ^
-DCUDNN_HOME="C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v9.0" ^
-Dtensorflow_WIN_CPU_SIMD_OPTIONS=/arch:AVX `

Everything looks good until the compiler_arch_native_support failed.
It manage to configure the build so I continue with: 
MSBuild /p:Configuration=Release tf_python_build_pip_package.vcxproj 

The last 9999 line of the failed log is here. 
[tf-compile-build-error-log.txt](https://github.com/tensorflow/tensorflow/files/1484750/tf-compile-build-error-log.txt)

Warmest Regards,
Colman

**Comment by Carmezim** [comment]
AFAIK CUDA 9 and cuDNN 7 aren't supported on Windows yet.

**Comment by mrry** [comment]
@colmantse It looks like something is failing in the RE2 build, apparently in test code. Since we don't need that code, I've created PR #14696 to disable it. Could you please try patching it in (or making the one-line change to `tensorflow/contrib/cmake/external/re2.cmake`) and let us know if that fixes things?

**Comment by colmantse** [comment]
Thank you, by doing so the error is reduced from 90 to 1. It now gives:

```
"D:\projects\tensorflow\tensorflow\contrib\cmake\build\tf_python_build_pip_package.vcxproj" (default target) (1) ->
"D:\projects\tensorflow\tensorflow\contrib\cmake\build\pywrap_tensorflow_internal.vcxproj" (default target) (3) ->
"D:\projects\tensorflow\tensorflow\contrib\cmake\build\tf_core_gpu_kernels.vcxproj" (default target) (133) ->
(CustomBuild target) ->
  C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\VC\VCTargets\Microsoft.CppCommon.targets(17
1,5): error MSB6006: "cmd.exe" exited with code 1. [D:\projects\tensorflow\tensorflow\contrib\cmake\build\tf_core_gpu_k
ernels.vcxproj]

    2629 Warning(s)
    1 Error(s)

Time Elapsed 01:25:22.60
```

**Comment by colmantse** [comment]
Hi, thanks for the advice. I have implemented the change and the error reduced from 90 to 1 as follow:

```
"D:\projects\tensorflow\tensorflow\contrib\cmake\build\tf_python_build_pip_package.vcxproj" (default target) (1) ->
"D:\projects\tensorflow\tensorflow\contrib\cmake\build\pywrap_tensorflow_internal.vcxproj" (default target) (3) ->
"D:\projects\tensorflow\tensorflow\contrib\cmake\build\tf_core_gpu_kernels.vcxproj" (default target) (133) ->
(CustomBuild target) ->
  C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\VC\VCTargets\Microsoft.CppCommon.targets(17
1,5): error MSB6006: "cmd.exe" exited with code 1. [D:\projects\tensorflow\tensorflow\contrib\cmake\build\tf_core_gpu_k
ernels.vcxproj]

    2615 Warning(s)
    1 Error(s)

Time Elapsed 01:25:40.39
```

I googled and it seems this is a path problem, i rebuild the config using this but the problem persist. 
`cmake .. -A x64 -DCMAKE_BUILD_TYPE=Release ^
-DSWIG_EXECUTABLE="C:/swigwin-3.0.12/swig.exe" ^
-DPYTHON_EXECUTABLE="C:/ProgramData/Anaconda3/python.exe" ^
-DPYTHON_LIBRARIES="C:/ProgramData/Anaconda3/libs/python36.lib" ^
-DPYTHON_INCLUDE_DIR="C:/ProgramData/Anaconda3/Include" ^
-DNUMPY_INCLUDE_DIR="C:/ProgramData/Anaconda3/lib/site-packages/numpy/core/include" ^
-Dtensorflow_ENABLE_GPU=ON ^
-DCUDNN_HOME="C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v9.0" ^
-Dtensorflow_WIN_CPU_SIMD_OPTIONS=/arch:AVX`

here is the last 9999 of the new error log.
[tf-compile-build-error-log.txt](https://github.com/tensorflow/tensorflow/files/1485313/tf-compile-build-error-log.txt)

**Comment by colmantse** [comment]
I figure this might be vc 2017 cuda 9 problem, i switched back to vc 2015 and cuda 8 and finally successfully compile it and get it installed. It will still be great if the issue with vc 2017/cuda9 could be resolved tho. Thank you very much for your help.

**Comment by mrry** [comment]
There are a lot of warnings about code page issues. For example:

>  C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v9.0\include\cuda_runtime.h : warning C4819: The file contains a character that cannot be represented in the current code page (950). Save the file in Unicode format to prevent data loss (compiling source file D:\projects\tensorflow\tensorflow\core\kernels\slice_op.cc) [D:\projects\tensorflow\tensorflow\contrib\cmake\build\tf_core_kernels.vcxproj]

I'm not sure how the system code page interacts with the Visual C++ compiler&mdash;and there's no indication in the error message about exactly what character is causing problems&mdash;but it seems like it might be necessary to change your system settings to get this to build cleanly.

**Comment by colmantse** [comment]
thanks for the direction, may you clarify a little bit more on changing system settings? I am not very sure where should I be looking at. Also, is compiler_arch_native_support failed an issue? I noticed after the upgrade from tf 1.3 to 1.4, my gpu utility (2 1080 ti) drop from 90% to 30 %, efficiency decreased as well by more than 3 quarters. This is why I try to build from source ( I previously use pip install tensorflow_gpu --upgrade). I notice my pc support AVX2 but I am not sure how to compile arch:AVX/AVX2 together, but nevertheless, the new build does not yield any performance gain. CPU remains excessively high in usage (abnormally) and gpu outrageously low utility. I am pretty desperate now, would be great to have your advice.

**Comment by colmantse** [comment]
Additionally I am not able to compile with AVX2, despite when running tf 1.4 installed from pip, it says my pc supports so and would benefit from compiling with it. Below is the log, building was successful, but compiling yield below. Would be great if someone could help.
[compile AXV2 error.txt](https://github.com/tensorflow/tensorflow/files/1499033/compile.AXV2.error.txt)

**Comment by sylvain-bougnoux** [comment]
FYI I've been able to compile everything (without testing python bindings, and using AVX, not AVX2), see #14801. However I've discovered that despite everything ran fine, the VS2017 (until 15.4) distribution introduced a bug in the /WHOLEARCHIVE trick resulting in unfilled factories (session, device...) [as mentioned in](https://developercommunity.visualstudio.com/content/problem/132766/wholearchive-linker-flag-is-dysfunctional-since-15.html). Therefore I am stuck using vs2017 for linking my application with TF. @aluo-x

**Comment by colmantse** [comment]
thank you for the pointer!

**Comment by gunan** [comment]
If you are seeing a failure similar to this one:
http://ci.tensorflow.org/job/tf-master-win-gpu-cmake/

This looks like an nvcc bug, and it is reported to NVIDIA. We are waiting for their investigation

**Comment by colmantse** [comment]
thank you! however, I could not see the fail build log from your page. There seems a lot of experiment has been going on tho. Good to know!

**Comment by angersson** [comment]
@gunan Any updates on this? Can we link to an nvidia page / thread to help triage this issue?

**Comment by gunan** [comment]
https://bitbucket.org/eigen/eigen/pull-requests/351/win-nvcc/diff
Once the above change is merged, I will bump TF eigen dependency, and we should be all fixed.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by gunan** [comment]
Forgot to update, but TF master now builds with cuda9.0 and cudnn 7 without any problems.
We will not backport the fix to 1.4, as there were quite a bunch of fixes since 1.4 to fix the build for cuda9.

**Comment by CarltonSemple** [comment]
@gunan with cuda 9.0 & cudnn 7, I'm still seeing the error as of commit https://github.com/tensorflow/tensorflow/commit/25d275280dfb163674f81c7681c2c1d34545a155 . 

I used
```
cmake .. -A x64 -DCMAKE_BUILD_TYPE=Release ^
-DSWIG_EXECUTABLE=C:/tools/swigwin-3.0.12/swig.exe ^
-DPYTHON_EXECUTABLE=C:\Users\csemp\AppData\Local\Programs\Python\Python35\python.exe ^
-DPYTHON_LIBRARIES=C:\Users\csemp\AppData\Local\Programs\Python\Python35\libs\python35.lib ^
-Dtensorflow_ENABLE_GPU=ON ^
-DCUDNN_HOME="C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v9.0" ^
-Dtensorflow_WIN_CPU_SIMD_OPTIONS=/arch:AVX ^
-Dtensorflow_BUILD_CC_EXAMPLE=OFF
```

Additionally, I found the original error earlier in the logs. It's
```
 133>CustomBuild: (TargetId:8893)
                     CMake Error at tf_core_gpu_kernels_generated_adjust_contrast_op_gpu.cu.cc.obj.Release.cmake:222 (message): (TaskId:3323)
                       Error generating (TaskId:3323)
                       C:/Users/csemp/dev/tensorflowbuild/tensorflow/tensorflow/contrib/cmake/build/CMakeFiles/tf_core_gpu_kernels.dir/__/__/core/kernels/Release/tf_core_gpu_kernels_generated_adjust_contrast_op_gpu.cu.cc.obj (TaskId:3323)
                      (TaskId:3323)
                      (TaskId:3323)
```

I'm using MSBuild 15.5.180.51428, the one with Visual Studio 2017

**Comment by CarltonSemple** [comment]
@gunan https://github.com/tensorflow/tensorflow/commit/25d275280dfb163674f81c7681c2c1d34545a155 was from master as of yesterday

**Comment by gunan** [comment]
http://ci.tensorflow.org/job/tf-master-win-gpu-cmake/
Looks like our continuous build is healthy. But our build uses VS 2015,  and we do not set SIMD options.
So these may be the cause of the failure you are seeing.

The error message above is helpful, but still does not capture what compiler emits for that build target. Is it possible for you to file a new issue, and maybe share the full build output through pastebin?

**Comment by arunmandal53** [comment]
I have tried VS 2017 multiple time but failed to build so i did it with visual studio 2015. Hope we can build it using Visual Studio 2017 later soon. I have created a tutorial for building tensorflow on windows 10 using cmake and visual studio 2015 update 3 on my blog here http://www.python36.com/install-tensorflow-gpu-windows . Thanks @gunan for helping with great deep learning library.

**Comment by chaosink** [comment]
@arunmandal53 
Just change CUDA host compiler to v140 by the following CMake option (you can also find it in cmak-gui by checking "Advanced"):
```
-DCUDA_HOST_COMPILER="C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/bin/amd64/cl.exe
```
NonCUDA code can still be compiled by v141 in VS 2017.

Reference: https://medium.com/@shiweili/building-tensorflow-c-shared-library-on-windows-e79c90e23e6e

**Comment by arunmandal53** [comment]
@chaosink thank you. I will test it.
